go-controller/pkg/clustermanager/networkconnect/cluster_network_connect.go

```diff
@@ -0,0 +1,319 @@
+			}
+			for _, nad := range nads {
+				// check this NAD is controlled by a CUDN
+				controller := metav1.GetControllerOfNoCopy(nad)
+				isCUDN := controller != nil && controller.Kind == cudnGVK.Kind && controller.APIVersion == cudnGVK.GroupVersion().String()
+				if !isCUDN {
+					continue
+				}
+				network, err := util.ParseNADInfo(nad)
+				if err != nil {
+					return nil, nil, err
+				}
+				if !network.IsPrimaryNetwork() {
+					continue
+				}
+				// This NAD passed all validation checks, so it's selected by this CNC
+				nadKey := nad.Namespace + "/" + nad.Name
+				allMatchingNADKeys.Insert(nadKey)
+				discoveredNetworks = append(discoveredNetworks, &network)
+			}
+		case apitypes.PrimaryUserDefinedNetworks:
+			namespaceSelector, err := metav1.LabelSelectorAsSelector(&selector.PrimaryUserDefinedNetworkSelector.NamespaceSelector)
+			if err != nil {
+				return nil, nil, err
+			}
+			namespaces, err := c.namespaceLister.List(namespaceSelector)
+			if err != nil {
+				return nil, nil, err
+			}
+			for _, ns := range namespaces {
+				namespacePrimaryNetwork, err := c.networkManager.GetActiveNetworkForNamespace(ns.Name)
+				if err != nil {
+					return nil, nil, err
+				}
+				if namespacePrimaryNetwork.IsDefault() {
+					continue
+				}
+				// Get the NAD key for the primary network in this namespace.
+				// Since this is the PrimaryUserDefinedNetworks selector (for namespace-scoped UDNs),
+				// we expect exactly one NAD per network.
+				// Today we don't support multiple primary NADs for a namespace, so this is safe.
+				// Also note if the user misconfigures and ends up with CUDN and UDN for the same namespace,
+				// and if the CUDN was created first - which means the UDN won't be created successfully,
+				// then the user uses the P-UDN selector, the CUDN's NAD will be chosen here for this selector
+				// but that's a design flaw in the user's configuration, and expectation is for users to use
+				// the selectors correctly.
+				primaryNADs := namespacePrimaryNetwork.GetNADs()
+				if len(primaryNADs) != 1 {
+					return nil, nil, fmt.Errorf("expected exactly one primary NAD for namespace %s, got %d", ns.Name, len(primaryNADs))
+				}
+				// GetNADs() returns NADs in "namespace/name" format, so use directly
+				nadKey := primaryNADs[0]
+				allMatchingNADKeys.Insert(nadKey)
+				discoveredNetworks = append(discoveredNetworks, &namespacePrimaryNetwork)
+			}
+		default:
+			return nil, nil, fmt.Errorf("%w: unsupported network selection type %s", errConfig, selector.NetworkSelectionType)
+		}
+	}
+
+	return discoveredNetworks, allMatchingNADKeys, nil
+}
+
+func computeNetworkOwner(networkType string, networkID int) string {
+	return fmt.Sprintf("%s_%d", networkType, networkID)
+}
+
+// parseNetworkOwner extracts the topology type from an owner key.
+// Owner keys are formatted as "{topology}_{networkID}" (e.g., "layer3_1", "layer2_2").
+func parseNetworkOwner(owner string) (topologyType string, ok bool) {
```

Comment:
maybe call it `parseNetworkOwnerTopology`? I expected network name to be returned from the name :)